src/main/java/spoon/support/util/IdentifierHelper.java

```diff
@@ -0,0 +1,149 @@
+            "do", "goto", "private", "this",
+            "break", "implements", "protected", "throw",
+            "else", "import", "public", "throws",
+            "case", "enum", "instanceof", "return", "transient",
+            "catch", "extends", "try",
+            "final", "interface", "static",
+            "class", "finally", "strictfp", "volatile",
+            "const", "native", "super", "while", "_",
+            "true", "false", "null",
+            "void", "boolean", "int", "long", "float",
+            "double", "byte", "char", "short"
+    };
+
+    private static String[] structuralIdentifierExceptions = {
+            CtPackage.TOP_LEVEL_PACKAGE_NAME,
+            CtModule.TOP_LEVEL_MODULE_NAME,
+            CtType.NAME_UNKNOWN,
+            THIS_ARGUMENT
+    };
+
+    private static String[] otherReferenceIdentifierExceptions = {
+            CtExecutableReference.CONSTRUCTOR_NAME,
+            CtExecutableReference.UNKNOWN_TYPE,
+            CtTypeReference.NULL_TYPE_NAME,
+            Void.TYPE.getName(),
+            Boolean.TYPE.getName(),
+            Integer.TYPE.getName(),
+            Long.TYPE.getName(),
+            Float.TYPE.getName(),
+            Double.TYPE.getName(),
+            Byte.TYPE.getName(),
+            Character.TYPE.getName(),
+            Short.TYPE.getName(),
+            CLASS_SUFFIX,
+            NULL_TYPE_NAME
+    };
+
+    /**
+     * This method checks that the given identifier respects the definition
+     * given in the JLS (see: https://docs.oracle.com/javase/specs/jls/se9/html/jls-3.html#jls-3.8
+     *
+     * @param identifierType: allow to use different exception strategy depending on the type of element to check:
+     *                      if it's a {@link IdentifierType#STRUCTURAL_ELEMENTS} then all JLS rules must be respected, excepted for top level package and module names and anonymous types which are using figure for their name (see {@link CtClass#isAnonymous()});
+     *                      if it's a {@link IdentifierType#PACKAGE_REFERENCE} then dots are also accepted;
+     *                      if it's a {@link IdentifierType#OTHER_REFERENCE} then primitive types names, numbers and few other exceptions are accepted.
+     */
+    public static void checkIdentifier(String identifier, IdentifierType identifierType) {
+        String[] identifierExceptions;
+
+        // we check first against exception based on the identifier type
+        switch (identifierType) {
+            case STRUCTURAL_ELEMENTS:
+                identifierExceptions = structuralIdentifierExceptions;
+                break;
+
+            case OTHER_REFERENCE:
+                identifierExceptions = otherReferenceIdentifierExceptions;
+                break;
+
+            default:
+                identifierExceptions = new String[0];
+                break;
+        }
+
+        // in case of exception: it's accepted directly
+        for (String identifierException : identifierExceptions) {
```

Comment:
I suggest to create several static `HashSet<String>` - one for each identifier type - and then simply call `Set.contains(str)`. I would say it will be faster, then current algorithm.
WDYT?

Reply:
I'll change that.